File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(list, predicate)
  for _, x in ipairs(list) do
    if predicate(x) then 
      return string.lower(x)  
    end
  end
  return nil  
end


-- Write your powers generator here
function powers_generator(base, limit)
  return coroutine.create(function ()
    local exponent = 0
    local value = base ^ exponent
    while value <= limit do
      coroutine.yield(value)
      exponent = exponent + 1
      value = base ^ exponent
    end
  end)
end
-- Write your say function here
function say(word)
  if word == nil then
    return ""
  end
  return function(next)
    if next == nil then
      return word
    else
      return say(word .. " " .. next)
    end
  end
end
-- Write your line count function here
function meaningful_line_count(filename)
    local valid_lines = 0
    local file = io.open(filename, "r")

    if not file then
        error("No such file: " .. filename)
    end

    for line in file:lines() do
        local stripped_line = line:gsub("^%s*(.-)%s*$", "%1")
        if stripped_line ~= "" and stripped_line:sub(1, 1) ~= "#" then
            valid_lines = valid_lines + 1
        end
    end

    file:close()
    return valid_lines
end
-- Write your Quaternion table here

Quaternion = {}

quaternionmeta = {
  __index = {
  coefficients = function(self)
    return {self.a, self.b, self.c, self.d}
  end,
  conjugate = function(self)
    return Quaternion.new(self.a, -self.b, -self.c, -self.d)
  end
  },
  __add = function(self, other)
    return Quaternion.new(self.a + other.a,self.b + other.b,self.c + other.c,self.d + other.d)
  end,
  __mul = function(self, other)
    return Quaternion.new(self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d,
      self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c,
      self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b,
      self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a)
  end,
  __eq = function(self, other)
    return self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d
  end,
  __tostring = function(self)
    local function format_term(value, unit)
        if value == 0 then
            return ""
        elseif value == 1 then
            return unit
        elseif value == -1 then
            return  '-' .. unit
        else
            return tostring(value) .. unit
        end
    end

    local string_rep = {}
     
    if self.a ~= 0 then
        table.insert(string_rep, tostring(self.a))
    end
    
    

    table.insert(string_rep, format_term(self.b, 'i'))
    table.insert(string_rep, format_term(self.c, 'j'))
    table.insert(string_rep, format_term(self.d, 'k'))
    -- Filter out any empty strings
    local filtered_string_rep = {}
    for _, term in ipairs(string_rep) do
        if term ~= "" then
            table.insert(filtered_string_rep, term)
        end
    end

    string_rep = filtered_string_rep
    if #string_rep == 0 then
        return "0"
    end
    local ans = table.concat(string_rep, "+")
    ans = string.gsub(ans,"%+%-", "-")
    return ans
    --return '<' .. self.a .. ',' .. self.b .. ',' .. self.c .. ',' ..self.d .. '>'
  end
}

Quaternion.new = function (a,b,c,d)
  return setmetatable({a = a,b = b,c = c,d = d}, quaternionmeta)
end

File: python/exercises.py
from dataclasses import dataclass


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


def first_then_lower_case(list, predicate, /):
    for x in list:
        if predicate(x):
            return x.lower()
    return None


def powers_generator(*, base, limit):
    exponent = 0
    value = base ** exponent
    while value <= limit:
        yield value
        exponent += 1
        value = base ** exponent
        


def say(word=None):
    if word is None:
        return ""
    
    def next_word(next=None):
        if next is None:
            return word
        else:
            return say(f"{word} {next}")
    
    return next_word

    
def meaningful_line_count(filename, /):
    try:
        valid_lines = 0
        with open(filename, 'r', encoding='utf-8') as file:
            for line in file:
                stripped_line = line.strip()
                if stripped_line and stripped_line[0] != '#':
                    valid_lines += 1
        return valid_lines
    except FileNotFoundError:
        raise FileNotFoundError(f"No such file: {filename}")


@dataclass(frozen=True)
class Quaternion:
    a:float
    b:float
    c:float
    d:float

    def __init__(self, a, b, c, d, /):
        object.__setattr__(self, 'a', a)
        object.__setattr__(self, 'b', b)
        object.__setattr__(self, 'c', c)
        object.__setattr__(self, 'd', d)

    def __str__(self):
        def format_term(value, unit):
            if value == 0:
                return ""
            elif value == 1:
                return unit
            elif value == -1:
                return f"-{unit}"
            else:
                return f"{value}{unit}"

        string_rep = []
        if self.a != 0:
            string_rep.append(str(self.a))
        string_rep.append(format_term(self.b, 'i'))
        string_rep.append(format_term(self.c, 'j'))
        string_rep.append(format_term(self.d, 'k'))

        filtered_string_rep = []
        for term in string_rep:
            if term:
                filtered_string_rep.append(term)
        string_rep = filtered_string_rep
        
        if not string_rep:
            return "0"
        
        ans = "+".join(string_rep)
        return ans.replace("+-", "-")
    
    def __eq__(self, other):
        return self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d
    
    def __mul__(self,other):
        return Quaternion(self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d,
      self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c,
      self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b,
      self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a)

    def __add__(self, other):
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)
    
    @property
    def coefficients(self):
        return (self.a, self.b, self.c, self.d)
    
    @property
    def conjugate(self):
        return Quaternion(self.a, -self.b, -self.c, -self.d)

